lib/fetch.js

```diff
@@ -89,29 +93,54 @@ var fetchDirectory = function (state, callback) {
 };
 
 var fetchGit = function (state, callback) {
-    var exec = require('child_process').exec,
-        tmpPath = ospath.tmp() + '/' + uuid.v1(),
-        input = state.template.input,
-        command = 'git clone ' + input + ' ' + tmpPath;
+    var tmpPath = ospath.tmp() + '/' + uuid.v1(),
+        inputUrl = state.template.input.split('#')[0],
+        inputRef = state.template.input.split('#')[1];
 
     state.template.tmp = tmpPath;
 
-    exec(command, function (err) {
-        if (err) {
-            callback(err);
-            return;
-        }
-        if (isDirectory(tmpPath + '/.git')) {
-            fsExtra.remove(tmpPath + '/.git', function (err) {
+    async.series([
+        function (next) {
+            exec('git clone ' + inputUrl + ' ' + tmpPath, function (err) {
                 if (err) {
-                    callback(err);
+                    next(err);
                     return;
                 }
-                callback(null, state);
+                next(null);
             });
-        } else {
-            callback(null, state);
+        },
+        function (next) {
+            if (inputRef) {
+                exec('cd ' + tmpPath + ' && git reset --hard ' + inputRef, function (err) {
```

Comment:
This is cool!  Is `git reset --hard` necessary right after cloning?  There will be no working directory changes, so it might suffice to just `git checkout` the `inputRef`.